cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java

```diff
@@ -0,0 +1,114 @@
+import co.cask.cdap.api.schedule.Schedule;
+import co.cask.cdap.api.schedule.ScheduleSpecification;
+import co.cask.cdap.api.service.ServiceSpecification;
+import co.cask.cdap.api.spark.SparkSpecification;
+import co.cask.cdap.api.worker.WorkerSpecification;
+import co.cask.cdap.api.workflow.WorkflowSpecification;
+import co.cask.cdap.data2.metadata.dataset.MetadataDataset;
+import co.cask.cdap.data2.metadata.store.MetadataStore;
+import co.cask.cdap.internal.schedule.StreamSizeSchedule;
+import co.cask.cdap.internal.schedule.TimeSchedule;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A {@link AbstractSystemMetadataWriter} for an {@link Id.Application application}.
+ */
+public class AppSystemMetadataWriter extends AbstractSystemMetadataWriter {
+
+  private final ApplicationSpecification appSpec;
+
+  public AppSystemMetadataWriter(MetadataStore metadataStore, Id.Application entityId,
+                                 ApplicationSpecification appSpec) {
+    super(metadataStore, entityId);
+    this.appSpec = appSpec;
+  }
+
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    addPrograms(properties);
+    addSchedules(properties);
+    // appSpec.getPlugins() returns all instances of all plugins, so there may be duplicates.
+    // we only store unique plugins right now
+    Set<PluginClass> existing = new HashSet<>();
+    for (Plugin plugin : appSpec.getPlugins().values()) {
+      if (!existing.contains(plugin.getPluginClass())) {
+        addPlugin(plugin.getPluginClass(), null, properties);
+        existing.add(plugin.getPluginClass());
+      }
+    }
+    return properties.build();
+  }
+
+  @Override
+  String[] getSystemTagsToAdd() {
+    return new String[] {
+      appSpec.getName(),
+      appSpec.getArtifactId().getName()
+    };
+  }
+
+  private void addPrograms(ImmutableMap.Builder<String, String> properties) {
+    addPrograms(ProgramType.FLOW, appSpec.getFlows().values(), properties);
+    addPrograms(ProgramType.MAPREDUCE, appSpec.getMapReduce().values(), properties);
+    addPrograms(ProgramType.SERVICE, appSpec.getServices().values(), properties);
+    addPrograms(ProgramType.SPARK, appSpec.getSpark().values(), properties);
+    addPrograms(ProgramType.WORKER, appSpec.getWorkers().values(), properties);
+    addPrograms(ProgramType.WORKFLOW, appSpec.getWorkflows().values(), properties);
+  }
+
+  private void addPrograms(ProgramType programType, Iterable<? extends ProgramSpecification> specs,
+                           ImmutableMap.Builder<String, String> properties) {
+    for (ProgramSpecification spec : specs) {
+      properties.put(programType.getPrettyName() + MetadataDataset.KEYVALUE_SEPARATOR + spec.getName(),
```

Comment:
so now with this user can scope to program type too ?
I think we have a lot of different scoping now beside schema.
@poornachandra : What's your thought on so many scopes ?
Even if we go with this we should for sure document this.